File: GraffiTab_iOS_SDK/Classes/Tasks/Streamable/Comments/GTEditCommentTask.m
//
//  EditCommentTask.m
//  GraffiTab-iOS
//
//  Created by Georgi Christov on 29/03/2015.
//  Copyright (c) 2015 GraffiTab. All rights reserved.
//

#import "GTEditCommentTask.h"

@implementation GTEditCommentTask

- (void)editCommentWithId:(long)commentId text:(NSString *)text successBlock:(void (^)(GTResponseObject *))successBlock failureBlock:(void (^)(GTResponseObject *))failureBlock {
    self.sBlock = successBlock;
    self.fBlock = failureBlock;
    
    NSString *string = [GTRequestBuilder buildEditComment];
    
    NSMutableDictionary *params = [NSMutableDictionary dictionaryWithDictionary:@{JSON_REQ_COMMENT_ID:@(commentId),
                                                                                  JSON_REQ_GENERIC_TEXT:text}];
    
    [self executePostWithUrl:string parameters:params cachePolicy:NSURLRequestReloadIgnoringLocalCacheData success:^(AFHTTPRequestOperation *operation, id responseObject) {
        NSDictionary *responseJson = responseObject;
        
        [self parseJsonSuccess:responseJson];
    } failure:^(AFHTTPRequestOperation *operation, NSError *error) {
        NSHTTPURLResponse *response = operation.response;
        NSInteger statuscode = response.statusCode;
        
        [self parseJsonError:statuscode];
    }];
}

- (id)parseJsonSuccessObject:(NSDictionary *)json {
    GTComment *p = [[GTComment alloc] initFromJson:json[JSON_RESP_COMMENT_COMMENT]];
    
    return p;
}

@end

File: GraffiTab-iOS-SDK/Tasks/Streamables/Comments/GTEditCommentTask.swift
//
//  GTEditCommentTask.swift
//  GraffiTab-iOS-SDK
//
//  Created by Georgi Christov on 11/05/2016.
//  Copyright © 2016 GraffiTab. All rights reserved.
//

import UIKit
import Alamofire
import AlamofireObjectMapper
import ObjectMapper

class GTEditCommentTask : GTNetworkTask {
    
    func editComment(streamableId: Int, commentId: Int, text: String, successBlock: (response: GTResponseObject) -> Void, failureBlock: (response: GTResponseObject) -> Void) -> Request {
        self.sBlock = successBlock
        self.fBlock = failureBlock
        
        let url = GTRequestBuilder.buildEditCommentUrl(streamableId, commentId: commentId)
        
        let commentParams = ["text":text]
        let params = ["comment":commentParams]
        
        return request(.PUT, URLString: url, parameters: params, encoding: .JSON, completionHandler: { (response: Response<AnyObject, NSError>) -> Void in
            if (response.result.isFailure) {
                if (response.response == nil) {
                    self.parseJSONError(nil)
                }
                else {
                    self.parseJSONError(response)
                }
            }
            else {
                let resp = response.result.value
                
                self.parseJSONSuccess(resp!)
            }
        })
    }
    
    override func parseJSONSuccessObject(JSON: AnyObject) -> AnyObject {
        let item = Mapper<GTComment>().map(JSON["comment"])
        
        NSNotificationCenter.defaultCenter().postNotificationName(GTEvents.CommentChanged, object: nil, userInfo: ["comment" : item!])
        
        return item!
    }
}
